### DISCUSSION TITLE
Data file (.csv) calculation problem

### DISCUSSION THREAD

**Discussion by ontinglo** [jspsych/jsPsych#1991] [Q&A]
hi all

I want to calculate the average value of a column of data and save the average value in a new column in the .csv file. Like what is shown below:

Before
![Before](https://user-images.githubusercontent.com/68586579/125560437-6c24a050-b851-45b3-845c-483b5a4fd70b.jpg)

After
![After](https://user-images.githubusercontent.com/68586579/125560453-3d43542d-df13-4a40-b4f3-6bafe7785b47.jpg)

I don't know in where (now i write codes in jsPsych.init() and what codes should I add to achieve such purposes. What I have right now is like the following, apparently it does not work. could you please give me some advice?

thanks a lot!
```js
 jsPsych.init({
        timeline: timeline,
        preload_images: [all_StimSet, all_StimSetT],
        show_preload_progress_bar: true,
        show_progress_bar: true,
        minimum_valid_rt: 300,
        default_iti: 50,

        on_data_update: function(data) {
            var TrustA_data = jsPsych.data.get().select('Trial_Trust_Value').mean();
            T_data = {
                meannn: TrustA_data,
            }
            jsPsych.data.write(T_data);
        },

        exclusions: {
            min_width: 800,
            min_height: 600,
        },
        on_finish: function() {
            jsPsych.data.get().localSave('csv', 'pilot_v5_mydata_Murk.csv');
        },

    })
```

**Comment by becky-gilbert** [comment]
Hi @ontinglo, I wouldn't do this in the `on_data_update` function, because that function runs _every_ time information is saved to the jsPsych data, i.e. after each trial. Instead, I think you just need to do the mean calculation once, after all of the relevant trials have finished. Some options for this are in a [call-function trial](https://www.jspsych.org/plugins/jspsych-call-function/), in the [`on_timeline_finish` function](https://www.jspsych.org/overview/timeline/#timeline-start-and-finish-functions), which runs after a set of trials is finished, or at the very end of the experiment in the jsPsych.init [`on_finish` function](https://www.jspsych.org/overview/callbacks/#on_finish-experiment). 

I also wouldn't suggest using the `jsPsych.data.write` function to save the mean to the data - instead you can use [`jsPsych.data.addProperties`](https://www.jspsych.org/core_library/jspsych-data/#jspsychdataaddproperties), which will add the mean to the data for all trials, or [`jsPsych.data.get().addToLast`](https://www.jspsych.org/core_library/jspsych-data/#addtolast), which will add the mean to the last trial in the data collection.

Here are the steps for calculating the mean and saving it to the data:
1. Use the jsPsych data [filter function](https://www.jspsych.org/core_library/jspsych-data/#filter) to select the relevant trial data
2. Use the jsPsych data [select function](https://www.jspsych.org/core_library/jspsych-data/#select) to select the relevant data column from those trials
3. Use the jsPsych data [subset function](https://www.jspsych.org/core_library/jspsych-data/#subset) to remove any missing data (because you can't calculate a mean when some of the values you're averaging over are null, empty, non-numeric etc.)
4. Use the jsPsych data [mean function](https://www.jspsych.org/core_library/jspsych-data/#mean) to get the mean of the remaining values
5. Save this mean value to the data using either `jsPsych.data.addProperties` or `jsPsych.data.get().addToLast`

Here's an example of averaging the RT values from responses to a series of images and saving the mean RT to the data.
```js
var trial_procedure = {
  timeline: [fixation, trial],
  timeline_variables: trial_info,
  on_timeline_finish: function() {
    var all_rt_values = jsPsych.data.get().filter({trial_type:'image-keyboard-response'}).select('rt').subset(function(x){ return x !== null; });
    var rt_mean = all_rt_values.mean();
    jsPsych.data.get().addToLast({mean: rt_mean});
  }
};
```

In your case, I'm not sure that you can easily filter the trials with a trial trust value in your data. So I suggest adding a flag in the data for those trials, something like `data: {task_part: 'trial_trust'}`. Then you can do this somewhere in your experiment after all of the relevant trials are finished, e.g. in your experiment's `on_finish` function:
```js
on_finish: function() {
  var all_trial_trust_values = jsPsych.data.get().filter({task_part: 'trial_trust'}).select('Trial_Trust_Value').subset(function(x){ return x !== null; });
  var trial_trust_mean = all_trial_trust_values .mean();
  jsPsych.data.get().addToLast({trial_trust_mean: trial_trust_mean});
  jsPsych.data.get().localSave('csv', 'pilot_v5_mydata_Murk.csv');
}
```
I hope this helps!
